Dockerfile bikeshedding

ciaranmcnulty opened the pull request:
A few improvements, it probably makes sense to step through the commits to get a sense of each change

This:
  * Reformats and renames some things
  * Reorders some elements to improve layer caching
  * Uses mounts for composer+npm to avoid having those tools in final images
  * Improves caching when docker cache is invalidated by using (shared) cache mounts
  * Runs the tests as part of docker build (but retains the make targets for local)

ciaranmcnulty committed:
Use the workdir more

Sets the workdir earlier and removes /<workdir> prefixes where it is set

This will make a future workdir move easier

ciaranmcnulty committed:
Add an npm base image target

To be consistent with the composer approach

ciaranmcnulty committed:
Use bind mounts for package.json and composer.json

This is partially to support later changes, but we only need these (perhaps?) in the final development container, and we definitely don't want composer in the prod image.

There is no change to the caching behaviour; changes in these files invalidate the cache the same as if they were COPYed

ciaranmcnulty committed:
Add caches for npm and composer

The npm cache is nice to have, it speeds up the install locally when the packages.json has been modified

The composer cache is shared between stages (by id=composer) so as well as the above benefits, the later stages (e.g. production) have been primed by the prior stages without needing to explicitly copy the vendor folder

ciaranmcnulty committed:
Rename stages + copy composer from more sensible stage

ciaranmcnulty committed:
Reorder the stages to run composer on real target platform

This allows us to run the platform checks rather than having to re-do them in the final targets

Ocramius commented on Dockerfile:
```diff
@@ -62,52 +46,102 @@ RUN export DEBIAN_FRONTEND="noninteractive" \
+
+RUN vendor/bin/phpunit
+RUN vendor/bin/phpcs
+RUN vendor/bin/psalm
```
Mentioned previously: this is not an OK approach.

While I do indeed like having `docker build` as a validation tool, having this as part of the build process prevents being able to have clear CI targets explaining what/where it went wrong.

In addition to that, anyone trying to repair this docker image (should something go wrong) will first have to comment this, then build, then adjust, then uncomment and rebuild.

ciaranmcnulty replied:
I don't understand the last paragraph?

Ocramius replied:
Assuming something changes in a dependency (for example, re-tagging), then it is not possible to develop on top of this project, due to the build steps preventing having a functioning image.

ciaranmcnulty replied:
Ive reverted the CI changes, so tests are explicitly run there in separate steps again. This is now a last defence against shipping a 'bad' prod image

ciaranmcnulty committed:
Change COPY to use --link flag

This improves layer caching and has literally no downsides

ciaranmcnulty committed:
Always run the tests before producing the production image

Added a 'tested' stage that has successfully passed the tests, then made production depend on it. This is a final check before the prod image is made that we haven't somehow skipped CI

(Actually because CI doesn't run in ARM this will catch ARM breakages)

ciaranmcnulty committed:
Base dev composer dependencies on prod ones

This makes the layer of the composer install smaller - it's just adding files to vendor and doesn't need to add the ones used in prod

Ocramius reviewed (approved):
Good middle-ground between testing-in-build and testing-in-ci, thanks @ciaranmcnulty!
